src/compiler/types.ts

```diff
@@ -1,11 +1,26 @@
 namespace ts {
-
+    /**
+     * Type of objects whose values are all of the same type.
+     * The `in` and `for-in` operators can *not* be safely used,
+     * since `Object.prototype` may be modified by outside code.
+     */
     export interface MapLike<T> {
         [index: string]: T;
     }
 
-    export interface Map<T> extends MapLike<T> {
-        __mapBrand: any;
+    /** It's allowed to get/set into a map with numbers. However, when iterating, you may get strings back due to the shim being an ordinary object (which only allows string keys). */
+    export type MapKey = string | number;

```

Comment:
My [test](https://github.com/andy-ms/perf/blob/master/numberMap.js) shows a number-keyed Map to be slightly more efficient.